### ISSUE TITLE
Consumers: exceptions in one handler cause others for the same topic suffix to be skipped

### ISSUE THREAD

**Issue by nphilipp** [fedora-infra/bodhi#3422]
@bowlofeggs @cverna @abompard @pypingou sorry for the spam, but I guess this is best discussed in an issue.

Before the automatic update handler, all handlers listened for different message topic suffixes. This one however listens to the same suffix as the signed handler, meaning that an exception in the latter would cause the automatic update handler to be skipped, see the relevant code in `bodhi/server/consumers/__init__.py`:

``` python
    def __call__(self, msg: fedora_messaging.api.Message):  # noqa: D401
        [...]
        try:
            if msg.topic.endswith('.buildsys.tag'):
                log.debug('Passing message to the Signed handler')
                self.signed_handler(msg)

                log.debug('Passing message to the Automatic Update handler')
                self.automatic_update_handler(msg)

            if msg.topic.endswith('.greenwave.decision.update'):
                log.debug('Passing message to the Greenwave handler')
                self.greenwave_handler(msg)
        except Exception as e:
            error_msg = f'{str(e)}: Unable to handle message: {msg}'
            log.exception(error_msg)
            raise fedora_messaging.exceptions.Nack(error_msg)
```

I'm not sure if doing the `try: ... except: ...` dance for each suffix/handler combination individually is the right thing or not...

Right now:
* If `signed` raises an exception, `automatic_update` gets skipped and by way of the raised `Nack`, the message gets put into the queue again.
* If `automatic_updates` raises an exception, the message will be put into the queue (per `Nack`) even though the `signed` handler has run successfully. This makes me cross my fingers that the `signed` handler is idempotent. :wink:

The alternative I have in mind re: handling exceptions for each handler individually would change this to:
* If any handler throws an exception, this is caught, stored, but the other handlers are still processed.
* When all handlers have been processed, if any one of them failed the `Nack` is raised to put the message back into the queue. This would require that all our handlers are idempotent, i.e. can deal with being called multiple times on the same message, recognizing that this is the case and simply doing nothing when called a 2nd, 3rd, … time on the same message when they processed it successfully before.

I've created draft PR #3421 to illustrate what I have in mind.

What do you think?

**Comment by abompard** [comment]
A couple questions / remarks :
1. Do we actually want to process the automatic_update handler if the signed handler fails?
2. If we do, then yeah storing the failure in a variable and raising Nack if it is true seems like a good solution
3. All handlers should be either idempotent or handle duplicate messages, because the bus may send messages multiples times (in case of network failure when the ack should have been sent by the client for example).

**Comment by pypingou** [comment]
> Do we actually want to process the automatic_update handler if the signed handler fails?

That's likely, the later one "just" marks the updates as being signed. If for some reason it fails, we still want to create the rawhide update.

**Comment by abompard** [comment]
But then, the update will contain unsigned packages?

**Comment by pypingou** [comment]
> But then, the update will contain unsigned packages?

Bodhi doesn't sign the packages, it just has a small indication in its UI whether the builds of an update have been signed or not. It helps the package know where their packages lay.

**Comment by abompard** [comment]
If it's just a minor UI update then I guess yeah it's OK to keep processing handlers.

**Comment by nphilipp** [comment]
@abompard the indication in the UI is already there, so no change in UI involved. With this, I'll look at how well the signed & automatic update handlers cope with receiving and processing the same message twice.

**Comment by abompard** [comment]
Yeah I understand, what I meant is that if the consequence of crashing on the signed handler is just that a minor part of the UI is not going to be updated, it's not worth blocking the automatic handler on that.
So the change seems good to me in that regard.

**Comment by bowlofeggs** [comment]
On Mon, 2019-07-29 at 03:04 -0700, Pierre-Yves Chibon wrote:
> it just has a small indication in its UI whether the builds of an
> update have been signed or not.

This is not correct - builds being signed is part of Bodhi's state
machine. The UI thing is just a nice thing to have, but the field is
for the state machine and having the data be correct there is important
or the composer will crash.
